### PR TITLE
Fixing filebeat holding onto old file descriptors

### PR DESCRIPTION
**Pull request by al3jandrosg**
Under certain scenariosFilebeat may hold onto old
log file descriptors. There are various settings
on Filebeat to prevent this from happening. This
patch adds to the charm two key settings that
help prevent this issue:

- ignore_older
- clean_inactive

Note that while these configuration parameters used to not be specified in the filebeat config files, they will now always be rendered. The charm configuration however aligns with the upstream defaults so this is not expected to alter behavior of existing deployments upon upgrading.

Fixing: [#87](https://github.com/canonical/layer-filebeat/issues/87)

### CODE REVIEW

File: config.yaml
Comment (aieri) [inline config.yaml:1]: Reading through the documentation, it seems that these settings must be carefully tuned and are not totally independent. E.g.
> The clean_inactive setting must be greater than ignore_older +
scan_frequency to make sure that no states are removed while a file is still being harvested.

It would seem to me safer and more in the spirit of charms to only expose the `clean_inactive` setting and automatically calculate sane values for the related settings (and reject input that doesn't make sense)

File: config.yaml
Comment (aieri) [inline config.yaml:65]: the related bug only mentions `clean_inactive`, why do we need to expose all three settings?

File: templates/filebeat-5.yml
Comment (aieri) [inline templates/filebeat-5.yml:21]: can we do some sanity checking of these values? Passing them verbatim will lead to breakage if the user configures them incorrectly

File: config.yaml
Comment (aieri) [reply config.yaml:1]: another option would be making this a boolean, or a switch having a limited number of presets, e.g.
```
forget_inactive:
    type: string
    default: "never"
    description: |
        Make Filebeat forget about files that have not been updated in a while. Possible values: never|conservative|aggressive.
        [explain the values here]
        WARNING: this setting could lead to data loss if...
        WARNING: this setting could cause Filebeat to retransmit a file from the beginning if...
        See also <URL> for further details
```

**Review by aieri — CHANGES_REQUESTED** [review]

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:65]: I should remove `close_timeout` from the patch as it may not necessarily be useful and could cause problems. It was initially added targeting some specific scenarios but overall those are unlikely.

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:1]: I think we could modify the `ignore_older` charm default as it needs to be longer than `close_inactive` which defaults to `5m`. Hence, I suggest setting the default to `1h` so that we at least ensure logs modified within the last `1h` are collected.

 `clean_inactive` is disabled by default but we need to make sure users use a value larger than `ignore_older + scan_frequency`, thus I'll add sanity checks to the charm. Should we throw an error or a warning if the above rule is not satisfied? I believe a warning would suffice, as I would not expect anything to break (other than logs getting sent repeatedly and duplicated) and hence find myself reluctant to block a unit and throw an error. Thoughts?

Overall I believe that giving the users the flexibility (but still doing damage control with sanity checks) of configuring the values themselves would be preferred, rather than a predefined (and thus opinionated) number of presets. This would also prevent issues with unforeseen edge cases that could arise due to the commonality of odd configs and deployments we are used to finding in the wild.

File: templates/filebeat-5.yml
Comment (al3jandrosg) [reply templates/filebeat-5.yml:21]: I guess we should block the unit and throw an error if a sanity check on config values fails. Thoughts?

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:65]: `ignore_older` was added as it resolves another issue reported (customer ticket) and allows users also to avoid collecting logs they don't need.

File: config.yaml
Comment (aieri) [reply config.yaml:1]: I think the fewer knobs we offer users, the harder it will be for them to shoot themselves in the foot. If profiles seem too restrictive, I would suggest offering one knob and calculating sane values for the dependent ones. For example, allow users to tune `clean_inactive`, but calculate `ignore_older` and `scan_frequency`.

File: config.yaml
Comment (aieri) [reply config.yaml:65]: logs that aren't needed can be ignored via the `excluded_files` charm option; in any case I would prefer to keep the two PRs separate

File: templates/filebeat-5.yml
Comment (aieri) [reply templates/filebeat-5.yml:21]: yes, setting the charm to blocked until the configuration is amended seems reasonable to me

Comment (al3jandrosg) [reply templates/filebeat-5.yml]: I have resubmitted the patch with sanity checks for the values, the unit will get into blocked statem if the checks fail and will log extra details on why it failed.

File: config.yaml
Comment (al3jandrosg) [reply config.yaml]: just `clean_inactive` and `ignore_older` included in the patch as the issue is resolved by setting those, no need for `close_timeout`

**Comment by al3jandrosg** [comment]
@aieri any feedback on the latest changes by any chance? I am looking forward to pushing this forward as my customer is frequently having issues.

File: config.yaml
Comment (aieri) [reply config.yaml:1]: why do you need both clean_inactive and ignore_older?

File: config.yaml
Comment (aieri) [reply config.yaml]: why do you also need `ignore_older`? Isn't `excluded_files` sufficient for your use-case?

File: reactive/filebeat.py
Comment (aieri) [inline reactive/filebeat.py:196]: are you sure this prevents incorrect values from being rendered by the template? Incorrect values should never make it on disk

File: reactive/filebeat.py
Comment (aieri) [inline reactive/filebeat.py:211]: this is a bit obscure, would it be perhaps better to have separate log lines if the validation of either string fails? A small explanation of what "invalid" means would also be nice

**Review by aieri — CHANGES_REQUESTED** [review]

File: config.yaml
Comment (aieri) [inline config.yaml:65]: advice: you could also make your life a hell of a lot easier by only accepting int with the meaning of minutes. All of your validation functions would almost disappear

**Comment by aieri** [comment]
> @aieri any feedback on the latest changes by any chance? I am looking forward to pushing this forward as my customer is frequently having issues.

hey sorry for the lack of reviews. This PR was somehow still listed as waiting for your changes in our board, which meant nobody would have picked it up during their normal daily reviews

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:248]: done

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:211]: done

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:59]: makes sense to keep it disabled as it seems it's disabled by default

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:67]: makes sense to keep it disabled as it seems it's disabled by default

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:1]: added to the description

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:271]: indeed, that's unfortunate (but understandable with the charm in maintenance). I've tested all those cases I listed in the function and this works as expected though.

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:65]: correcting myself, I fully agree accepting only integers (in seconds for better granularity) simplifies the code quite a lot. Refactoring and testing

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:200]: This poses a challenge as we need `scan_frequency` to ensure compliance with `clean_inactive >= ignore_older + scan_frequency`. We could potentially try and read `scan_frequency` from the config and if does not exist then use the default (`10s`) for the calculation or assume a realistic arbitrary value (e.g., `scan_frequency = 10min`). The former is a bit cumbersome unless there's something in charmhelpers to help me there. Any suggestions @aieri ?

We could also add `scan_frequency` back as config option but I'm guessing you'd be reluctant to go that way (I don't know why that config got removed) hence why the two options above.

**Comment by aieri** [comment]
@al3jandrosg I would also add that although it's true that this charm does not have unit tests, I think your change should come with some (for its own functions). I have created #111 for this reason, so once that one is merged please rebase and add your tests.

File: reactive/filebeat.py
Comment (aieri) [reply reactive/filebeat.py:200]: `scan_frequency` was never an option so I don't think we should add it now if we don't actually need people to tune it. As we do know which versions of filebeat this charm will deploy, we also know what the default setting is. We can use that value for the sanity calculations (10s).
If you then want to be extra sure the payload will indeed use a 10s value, I'd be ok with explicitly setting it in the jinja config templates (hardcoded).

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:196]: Hey @aieri, I've been doing some testing and this does not work, no. If I change the config option values via `juju config` the sanity checks work as expected but if I pass the options with invalid values via the bundle the sanity checks never get executed. I've tried various workarounds and flags but nothing has worked. I'm awaiting feedback from Juju SMEs but any suggestions you may have would be very appreciated.

Comment (al3jandrosg) [reply reactive/filebeat.py]: this is now resolved, I've added a call to `sanity_check()` to the `render_filebeat_template()` function and if that fails the unit goes into blocked state. The config options are still getting rendered into the file, but in any case we'd need manual intervention to fix an invalid value so IMHO this should be fine.

File: reactive/filebeat.py
Comment (aieri) [inline reactive/filebeat.py:227]: suggestion: no casting needed if the config options are specified to be integers at the charm config level

File: reactive/filebeat.py
Comment (aieri) [inline reactive/filebeat.py:104]: please sanity check the config as early as possible. If you do it this late it cannot prevent the on-disk configuration from being rendered, meaning that an accidental restart of filebeat would cause it to load values that would fail the check

**Review by aieri — CHANGES_REQUESTED** [review]

File: config.yaml
Comment (aieri) [reply config.yaml]: @al3jandrosg this conversation is still open: why do you need to configure `ignore_older`? The original bug only ask for tuning `clean_inactive`

File: config.yaml
Comment (al3jandrosg) [reply config.yaml:53]: ah this is neat and simplifies things a lot, thanks

File: config.yaml
Comment (al3jandrosg) [reply config.yaml]: @aieri `ignore_older` is used by the customer(s) to avoid harvesting logs older than a certain time threshold. `exclude_files` is not precisely what they need. As for the file descriptor addressed in this PR, `ignore_older` helps the mitigation of potential issues using `clean_inactive`, as it ensures that no harvesters are triggered for older logs and the effects of `clean_inactive` i.e., removing the state of the file, are _safer_ (customers have raised concerns about using `close_*` and `clean_*` directives because the Filebeat docs warn about potential data loss side effects).

File: reactive/filebeat.py
Comment (aieri) [inline reactive/filebeat.py:201]: non-blocking: I suspect these `@when` are unnecessary since you're calling this function from within `render_filebeat_template()`

File: reactive/filebeat.py
Comment (aieri) [inline reactive/filebeat.py:226]: this should also be unnecessary given the restart at line 99

**Review by aieri — DISMISSED** [review]
there are a couple of things that can still be improved I think, but this is mergeable from my perspective (if unit tests will be provided at a later date as discussed)

File: reactive/filebeat.py
Comment (jneo8) [reply reactive/filebeat.py:201]: `@when("apt.installed.filebeat")` here will cause this function run twice. This should be remove.

File: reactive/filebeat.py
Comment (jneo8) [reply reactive/filebeat.py:226]: I suggest to remove this line because the function name `is_config_valid` should be a read-only check instead of raise side-effect(restart service).

**Review by jneo8 — DISMISSED** [review]

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:201]: agreed `@when("apt.installed.filebeat")` is not necessary, but isn't `@when("config.changed")` needed to call the function upon a `config.changed` hook or after changes to the config?  Unless the latter happens because `beat.render` is set upon any config change which would run `render_filebeat_template`, if this is the case then I guess we could live with the sanity checks only running when `render_filebeat_template` is called.

**Review by aieri — APPROVED** [review]

File: reactive/filebeat.py
Comment (chanchiwai-ray) [inline reactive/filebeat.py:200]: [non-blocking] I guess this is also not required. The beat-layers](https://github.com/canonical/layer-beats-base/blob/master/reactive/beats_base.py#L9-L12) has `config-changed` that `set_state("beat.render")`, and this trigger `render_filebeat_template()`

So this mean, `is_config_valid` can be a simple python function that check if the config is valid. Then, in the `render_filebeat_template()` we can have

```python
if not is_config_valid():
   status.blocked("Invalid config. Please check the logs for details.")
```

File: reactive/filebeat.py
Comment (chanchiwai-ray) [inline reactive/filebeat.py:81]: I cannot verify it now because the build is failing on master branch, but here it's return without removing the `beat.render` state. This might cause a continues loop on trying to run this function

File: reactive/filebeat.py
Comment (chanchiwai-ray) [reply reactive/filebeat.py]: `@when("config.changed")` is set in [beat-layers](https://github.com/canonical/layer-beats-base/blob/master/reactive/beats_base.py#L9-L12) and it also set `beat.render` flag on each `config_changed`

File: reactive/filebeat.py
Comment (aieri) [reply reactive/filebeat.py:81]: ah great find!

Comment (jneo8) [reply reactive/filebeat.py:81]: +1 to this one.

**Review by jneo8 — APPROVED** [review]
I am good once https://github.com/canonical/layer-filebeat/pull/109/files#r1689035311 resolved.

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py]: fixed, thanks

File: reactive/filebeat.py
Comment (al3jandrosg) [reply reactive/filebeat.py:83]: great find, fixed, thanks

**Review by chanchiwai-ray — APPROVED** [review]

**Review by aieri — APPROVED** [review]

**Comment by aieri** [comment]
force merging since we agreed that unit tests will be coming in a separate PR

**Pull request merged**
